### PR TITLE
[SYSTEMML-1677] Add a new 2D cross_entropy_layer

### PR DESCRIPTION
**Pull request by feihugis**
A 2D cross_entropy_layer would accept a tensor of shape (N,C,H,W), where the C axis contains scores for D classes, and output the average loss.

### CODE REVIEW

**Comment by dusenberrymw** [comment]
Thanks, will review tomorrow.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1664/

File: scripts/nn/layers/cross_entropy_loss2d.dml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * Cross-Entropy loss2d function.
```
Comment (dusenberrymw) [inline scripts/nn/layers/cross_entropy_loss2d.dml]: Let's call this "2D Cross-Entropy loss"

File: scripts/nn/layers/cross_entropy_loss2d.dml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+ source("nn/util.dml") as util
+ source("nn/layers/cross_entropy_loss.dml") as cross_entropy_loss
+
+forward = function(matrix[double] pred, matrix[double] y, int C)
+    return (double loss) {
+  /*
+   * Computes the forward pass for a cross-entropy loss2d function.  The
```
Comment (dusenberrymw) [inline scripts/nn/layers/cross_entropy_loss2d.dml]: "...for a 2D cross-entropy loss..."

File: scripts/nn/layers/cross_entropy_loss2d.dml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+   * inputs consist of N examples with a shape (Hin, Win), each point in
+   * the example with C dimensions corresponding to normalized 
+   * probabilities of C classes.
```
Comment (dusenberrymw) [inline scripts/nn/layers/cross_entropy_loss2d.dml]: Can you add a note that the loss is applied to each pixel location, and then averaged across all pixels and all examples?

File: scripts/nn/layers/cross_entropy_loss2d.dml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+   *
+   *   ```
+   *   L_i = -y_i^T * log(pred_i)
+   *   L = (1/N) sum(L_i) for i=1 to N
+   *   ```
+   *
+   * In these equations, `L` is the total loss, `L_i` is the loss for
+   * example `i`, `y_i` is the K-dimensional vector of target class
+   * probabilities, `pred_i` is K-dimensional vector of predicted
+   * class probabilities, and `N` is the number of examples.
```
Comment (dusenberrymw) [inline scripts/nn/layers/cross_entropy_loss2d.dml]: Can you update this section and the equations so that `L_ijk` is the loss for the pixel at location `(j, k)` in example `i`, and `L = (1/(N*H*W)) sum(L_ijk) for i=1 to N, j=1 to H, k=1 to W` so that it is clear that the returned loss is averaged across all pixel locations and all examples?

File: scripts/nn/test/grad_check.dml [hunk 5]
```diff
@@ -949,6 +951,48 @@ cross_entropy_loss = function() {
...
 
+cross_entropy_loss2d = function() {
+  /*
+   * Gradient check for the cross-entropy loss function.
+   */
+  print("Grad checking the cross-entropy loss2d function.")
+
+  # Generate data
+  N = 3 # num examples
+  C = 10 # num targets
```
Comment (dusenberrymw) [inline scripts/nn/test/grad_check.dml]: small: can you add an extra space to each of these lines so that there are two spaces before the `#` instead of one?  Python style requires two spaces at the end of the line before any inline comments.

**Comment by feihugis** [comment]
@dusenberrymw Sure, will work on the unit test now.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1675/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1676/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1686/

File: scripts/nn/test/test.dml [hunk 6]
```diff
@@ -410,6 +411,64 @@ cross_entropy_loss = function() {
   }
 }
 
+cross_entropy_loss2d = function() {
+  /*
+   * Test for the 2D cross-entropy loss function.
+   */
+  print("Testing the forward part of the 2D cross-entropy loss function")
```
Comment (dusenberrymw) [inline scripts/nn/test/test.dml]: Can you update this to `"Testing the 2D cross-entropy loss function."`?

File: scripts/nn/test/test.dml [hunk 7]
```diff
@@ -410,6 +411,64 @@ cross_entropy_loss = function() {
...
+              0 1 0
+              0 0 1
+              0 0 1
+              0 0 1", rows=N, cols=C*Hin*Win)
+
+  loss = cross_entropy_loss2d::forward(pred, y, C)
+
+  if (abs(loss - loss_expected) > 1.0e-7) {
+    print("ERROR: The 2D cross-entropy loss function ouptut (" + loss + ") does not match the expected value " + loss_expected + ".")
+  }
```
Comment (dusenberrymw) [inline scripts/nn/test/test.dml]: For testing consistency, can we use the `test_util::check_rel_error` function instead here, probably with `1e-3` & `1e-4` for the error and warning threshold arguments like the other tests?

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1690/

File: scripts/nn/test/grad_check.dml [hunk 8]
```diff
@@ -951,6 +952,45 @@ cross_entropy_loss = function() {
...
+  /*
+   * Gradient check for the 2D cross-entropy loss function.
+   */
+  print("Grad checking the 2D cross-entropy loss function.")
+
+  # Generate data
+  N = 3  # num examples
+  C = 10  # num targets
+  Hin = 28  # example height
+  Win = 28  # example width
```
Comment (dusenberrymw) [inline scripts/nn/test/grad_check.dml]: Using 28 was great while creating this PR to ensure that there was no stability issues, but I will reduce it to 5 when I merge just to reduce the amount of time it takes to run the test.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1708/
